src/metabase/api/table.clj

```diff
@@ -78,6 +68,34 @@
         (sync/sync-table! updated-table))
       updated-table)))
 
+(api/defendpoint PUT "/:id"
+  "Update `Table` with ID."
+  [id :as {{:keys [display_name entity_type visibility_type description caveats points_of_interest
+                   show_in_getting_started], :as body} :body}]
+  {display_name            (s/maybe su/NonBlankString)
+   entity_type             (s/maybe su/EntityTypeKeywordOrString)
+   visibility_type         (s/maybe TableVisibilityType)
+   description             (s/maybe su/NonBlankString)
+   caveats                 (s/maybe su/NonBlankString)
+   points_of_interest      (s/maybe su/NonBlankString)
+   show_in_getting_started (s/maybe s/Bool)}
+  (update-table id body))
+
+(api/defendpoint PUT "/bulk"
+  "Update all `Table` in `ids`."
+  [:as {{:keys [ids display_name entity_type visibility_type description caveats points_of_interest
+            show_in_getting_started], :as body} :body}]
+  {ids                     (s/pred vector?)
+   display_name            (s/maybe su/NonBlankString)
+   entity_type             (s/maybe su/EntityTypeKeywordOrString)
+   visibility_type         (s/maybe TableVisibilityType)
+   description             (s/maybe su/NonBlankString)
+   caveats                 (s/maybe su/NonBlankString)
+   points_of_interest      (s/maybe su/NonBlankString)
+   show_in_getting_started (s/maybe s/Bool)}
+  (map #(update-table % body) ids))
```

Comment:
@camsaul what do you think about this API? I wasn't sure if we had existing bulk-update patterns defined anywhere. I think it could also just be a single-purpose endpoint to bulk update `visibility_type` if this feels awkward.  I'll add some BE tests around it if you think it works as-is.

Reply:
We don't really have any bulk update patterns right now but seeing how we have things like `GET /whatever/` to fetch many vs `GET /whatever/:id` to fetch one it seems like `PUT /whatever` for bulk would be a good parallel for the existing `PUT /whatever/:id` for updating a single object

Reply:
This whole thing should be done in a transaction to so if updating one Table fails the changes to other Tables aren't persisted.

```clj
(db/transaction
  (mapv #(update-table % body) ids))
```

You should use `mapv` rather than `map` so the whole thing is done inside of the transaction instead of `map` which will return a lazy sequence that doesn't actually get realized until we start serializing the results to JSON

Reply:
I was thinking about this a bit more and it's probably better to rework `update-table!` to be an `update-tables!` function that efficiently updates all the Tables in a single app DB call rather than 100 DB calls for 100 tables (etc.)

```clj
(defn- update-tables!
  "Update properties of Tables with `ids`. Returns sequence of updated Tables in same order as `ids`."
  [ids {:keys [display_name entity_type visibility_type description caveats points_of_interest show_in_getting_started]
        :as   body}]
  (doseq [id ids]
    (api/write-check Table id))
  (let [id->original-visibility-type (db/select-id->field :visibility_type Table :id [:in (set ids)])]
    (api/check-500
     (m/mapply
      db/update-where! Table {:id [:in (set ids)]}
      ;; always update visibility type; update display_name, show_in_getting_started, entity_type if non-nil;
      ;; update description and related fields if passed in
      (assoc (u/select-keys-when body
               :non-nil [:display_name :show_in_getting_started :entity_type]
               :present [:description :caveats :points_of_interest])
             :visibility_type visibility_type)))
    (let [id->table (u/key-by :id (db/select Table :id [:in (set ids)]))]
      (doall
       (for [id ids]
         (let [updated-table   (id->table id)
               now-visible?    (nil? (:visibility_type updated-table)) ; only Tables with `nil` visibility type are visible
               was-visible?    (nil? (id->original-visibility-type id))
               became-visible? (and now-visible? (not was-visible?))]
           (when became-visible?
             (log/info (u/format-color 'green (trs "Table ''{0}'' is now visible. Resyncing." (:name updated-table))))
             (sync/sync-table! updated-table))
           updated-table))))))
```

Then the existing `PUT /api/table/:id` endpoint can call that for a single Table:

```clj
(api/defendpoint PUT "/:id"
  "Update `Table` with ID."
  [id
   :as {{:keys [display_name entity_type visibility_type description caveats points_of_interest show_in_getting_started]
         :as body} :body}]
  {display_name            (s/maybe su/NonBlankString)
   entity_type             (s/maybe su/EntityTypeKeywordOrString)
   visibility_type         (s/maybe TableVisibilityType)
   description             (s/maybe su/NonBlankString)
   caveats                 (s/maybe su/NonBlankString)
   points_of_interest      (s/maybe su/NonBlankString)
   show_in_getting_started (s/maybe s/Bool)}
  (update-tables! [id] body))
```

Reply:
I agree that `update-tables!` is much better. I didn't want to spend more time mucking around in Clojure when we're FE-constrained, so I left it and added a note indicating that a perf boost would be good.